contributed_definitions/nyaml/NXimage_set.yaml

```diff
@@ -1,38 +1,367 @@
+    magnitude(NX_NUMBER):
+      doc: |
+        Magnitude of the image intensity.
+      unit: NX_UNITLESS
+      dim: (n_i,)
+    axis_i(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along fastest direction.
+      unit: NX_LENGTH
+      dim: (n_i,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along fastest direction.
+
+  image_twod(NXdata):
+    doc: |
+      Two-dimensional image.
+    real(NX_NUMBER):
+      doc: |
+        Real part of the image intensity per pixel.
+      unit: NX_UNITLESS
+      dim: (n_j, n_i)
+    imag(NX_NUMBER):
+      doc: |
+        Imaginary part of the image intensity per pixel.
+      unit: NX_UNITLESS
+      dim: (n_j, n_i)
+    magnitude(NX_NUMBER):
+      doc: |
+        Magnitude of the image intensity.
+      unit: NX_UNITLESS
+      dim: (n_j, n_i)
+    intensity(NX_COMPLEX):
+      doc: |
+        Image intensity as a complex number as an alternative
+        to real and imag fields if values are stored as interleaved
+        complex numbers.
+      unit: NX_UNITLESS
+      dim: (n_j, n_i)
+    axis_j(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along fast direction.
+      unit: NX_LENGTH
+      dim: (n_j,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along fast direction.
+    axis_i(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along fastest direction.
+      unit: NX_LENGTH
+      dim: (n_i,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along fastest direction.
+
+  image_threed(NXdata):
+    doc: |
+      Three-dimensional image.
+    real(NX_NUMBER):
+      doc: |
+        Real part of the image intensity per pixel.
+      unit: NX_UNITLESS
+      dim: (n_k, n_j, n_i)
+    imag(NX_NUMBER):
+      doc: |
+        Imaginary part of the image intensity per pixel.
+      unit: NX_UNITLESS
+      dim: (n_k, n_j, n_i)
+    magnitude(NX_NUMBER):
+      doc: |
+        Magnitude of the image intensity.
+      unit: NX_UNITLESS
+      dim: (n_k, n_j, n_i)
+    intensity(NX_COMPLEX):
+      doc: |
+        Image intensity as a complex number as an alternative
+        to real and imag fields if values are stored as interleaved
+        complex numbers.
+      unit: NX_UNITLESS
+      dim: (n_k, n_j, n_i)
+    axis_k(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along slow direction.
+      unit: NX_LENGTH
+      dim: (n_k,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along slow direction.
+    axis_j(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along fast direction.
+      unit: NX_LENGTH
+      dim: (n_j,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along fast direction.
+    axis_i(NX_NUMBER):
+      doc: |
+        Pixel coordinate center along fastest direction.
+      unit: NX_LENGTH
+      dim: (n_i,)
+      \@long_name(NX_CHAR):
+        doc: |
+          Coordinate along fastest direction.
+
+  stack_oned(NXdata):
```

Comment:
Why do you make a distinction between a single `image_oned` and a stack of them. Can you not just store them in the same data object and then for a single image and set `n_img` to 1? Or would that not work dimension-wise?

The same is probably true for distinctions between `image_oned` and `image_twod`. If you loosen the restrictions on the dimensions, you could unify almost all these concepts (i..e, you need just one NXdata object), which would make this base class more succint. But, you loose the possibility of actually checking the dimensions, so I amnot sure if that's worth it.

Reply:
In MPES, we sort of went the other direction in the [/NXmpes/ENTRY/data group](https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXmpes.html#nxmpes-entry-data-group), leaving the dimensions intentionally vague so that any number of multi-dimensional data could be written.

Reply:
Enabling setting of constraints is the main reason as you discussed in your answer. The reason to distinguish between NXdata instance with an n_img or not is exactly for the purpose of avoiding that one is required to promote a single say 2d image to a de facto 3D array with the slow dimension just for the sake of communicating hey it is one image as you thought.

Reply:
Yeah, I get the point, if you explicitly want to restrict the dimensions in the base class with symbols, you sort of have to go that route..